x-pack/dockerlogbeat/pipereader/reader.go

```diff
@@ -0,0 +1,114 @@
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"syscall"
+
+	"github.com/containerd/fifo"
+	"github.com/docker/engine/api/types/plugins/logdriver"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+)
+
+// PipeReader reads from the FIFO pipe we get from the docker container
+type PipeReader struct {
+	fifoPipe    io.ReadCloser
+	byteOrder   binary.ByteOrder
+	lenFrameBuf []byte
+	bodyBuf     []byte
+	maxSize     int
+}
+
+// NewReaderFromPath creates a new FIFO pipe reader from a docker log pipe location
+func NewReaderFromPath(file string) (*PipeReader, error) {
+	inputFile, err := fifo.OpenFifo(context.Background(), file, syscall.O_RDONLY, 0700)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error opening logger fifo: %q", file)
+	}
+
+	return &PipeReader{fifoPipe: inputFile, byteOrder: binary.BigEndian, lenFrameBuf: make([]byte, 4), bodyBuf: make([]byte, 4*1024), maxSize: 2e6}, nil
+}
+
+// NewReaderFromReadCloser creates a new FIFO pipe reader from an existing ReadCloser
+func NewReaderFromReadCloser(pipe io.ReadCloser) (*PipeReader, error) {
+	return &PipeReader{fifoPipe: pipe, byteOrder: binary.BigEndian, lenFrameBuf: make([]byte, 4), bodyBuf: make([]byte, 4*1024), maxSize: 2e6}, nil
+}
+
+// ReadMessage reads a log message from the pipe
+// The message stream consists of a 4-byte length frame and a message body
+// There's three logical paths for this code to take:
+// 1) If length <0, we have bad data, and we cycle through the frames until we get a valid length.
+// 2) If length is valid but larger than the max buffer size, we disregard length bytes and continue
+// 3) If length is valid and we can consume everything into the buffer, continue.
+func (reader *PipeReader) ReadMessage(log *logdriver.LogEntry) error {
+	// loop until we're at a valid state and ready to read a message body
+	var lenFrame int
+	var err error
+	for {
+		lenFrame, err = reader.getValidLengthFrame()
+		if err != nil {
+			return errors.Wrap(err, "error getting length frame")
+		}
+		fmt.Fprintf(os.Stderr, "Got header specifying length %d\n", lenFrame)
+		if lenFrame <= reader.maxSize {
+			break
+		}
+
+		// 2) we have a too-large message. Disregard length bytes
+		_, err = io.CopyBuffer(nil, io.LimitReader(reader.fifoPipe, int64(lenFrame)), reader.bodyBuf)
+		if err != nil {
+			return errors.Wrap(err, "error emptying buffer")
+		}
+	}
+
+	//proceed with 3)
+	reader.setBuffer(lenFrame)
+	_, err = io.ReadFull(reader.fifoPipe, reader.bodyBuf[:lenFrame])
+	if err != nil {
+		return errors.Wrap(err, "error reading buffer")
+	}
+	fmt.Fprintf(os.Stderr, "Buffer now contains %#v\n", reader.bodyBuf[:lenFrame])
+	return proto.Unmarshal(reader.bodyBuf[:lenFrame], log)
+
+}
+
+// setBuffer checks the size and only reallocates if we need to
+func (reader *PipeReader) setBuffer(sz int) {
+	const minSz = 4 * 1024 // min 4KB read buffer
+
+	if len(reader.bodyBuf) >= sz {
+		return
+	}
+
+	if sz < minSz {
+		sz = minSz
+	}
+
+	reader.bodyBuf = make([]byte, sz)
+}
```

Comment:
hmm... this reader is per container right?

The default min buffer size is 4KB. for 100 containers we have a buffer of ~400KB. Quite low. But the max buffer size is ~2MB. With 100 containers we can end up with a read buffer of ~200MB (very worst case scenario). A compromise here might be a cutoff size that allocates if `sz >= cutoffSize`. This gives the runtime a chance to clear large allocations. A variation of this is a set of `sync.Pool` for different size classes (but maybe overkill for now).

e.g. 

```
func (reader *PipeReader) readBuffer(sz int) []byte {
  const minSize = 4*1024
  const cutoffSize = 32*1024

  if len(reader.bodyBuf) >= sz {
    return reader.bodyBuf[:sz]
  }

  if sz > cutoffSize {
    return make([]byte, sz)
  }

  reader.bodyBuf = make([]byte, sz)
  return reader.bodyBuf
}
```